multipart: send content-length in POST requests

vadmeste opened the pull request:
Send Content-Lenght header in POST requests to satisfy some proxies
(like squid) which require that header to work.

harshavardhana commented on api/src/main/java/io/minio/MinioClient.java:
```diff
@@ -1019,7 +1011,7 @@ private HttpResponse executeDelete(String bucketName, String objectName, Map<Str
     HttpResponse response = execute(Method.DELETE, BucketRegionCache.INSTANCE.region(bucketName),
                                     bucketName, objectName, null,
-                                    queryParamMap, null, null, 0);
+                                    queryParamMap, null, null, -1);
```
this might not be true for 'executeDelete()' when we implement Multi object delete API.  Change this to executeDelete() itself taking the content length.

vadmeste replied:
> this might not be true for 'executeDelete()' when we implement Multi object delete API. Change this to executeDelete() itself taking the content length.

Multi Delete is a POST and not DELETE http operation. I think we should keep it like that until we implement the new multi delete API.

harshavardhana replied:
ah correct.

balamurugana commented:
@vadmeste The fix should be at https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/MinioClient.java#L2543. Instead of sending `""`, we would need to send `new byte[0]`

balamurugana commented on api/src/main/java/io/minio/MinioClient.java:
```diff
@@ -671,15 +671,7 @@ public MediaType contentType() {
 
```
This is not a right fix.
`requestBody` is created only if `body != null`.  As per `createRequest()`, `body` is any object or `null`.  If `body` is non-null and other than `(body instanceof InputStream || body instanceof RandomAccessFile || body instanceof byte[])`, we won't need to compute md5sum and need to send `Transfer-Encoding: chunked`

vadmeste replied:
@balamurugana, if you focus on the function contentLength() itself, you can notice that the only change I've made is to return 0 instead of -1 when length == 0. Otherwise the behavior will be still the same.

balamurugana commented on api/src/main/java/io/minio/MinioClient.java:
```diff
@@ -1040,9 +1032,16 @@ private HttpResponse executePost(String bucketName, String objectName, Map<Strin
+    }
+    if (data instanceof CompleteMultipartUpload) {
+      len = ((CompleteMultipartUpload) data).toString().length();
+    }
```
Having execute{Head,Get,Post,Put,Delete} are just wrappers to execute() for convenience and readability across APIs  Spreading the logic of type checking will lead to confusion in future and its not a good idea.

balamurugana commented on api/src/main/java/io/minio/MinioClient.java:
```diff
@@ -1875,7 +1874,7 @@ public void makeBucket(String bucketName, String region)
     }
 
-    executePut(bucketName, null, headerMap, null, US_EAST_1, configString, 0);
+    executePut(bucketName, null, headerMap, null, US_EAST_1, configString, configString.length());
```
`String.length()` != `[].byte.length`.

We would need to convert string into UTF8 encoded byte array.

vadmeste replied:
This is for creating bucket, do we really need to convert to UTF8 ?

As @balamurugana, suggested, we need to enhance createRequest for a simpler logic.

harshavardhana replied:
it is largely `configString.getBytes("UTF-8").length()`

vadmeste replied:
> it is largely configString.getBytes("UTF-8").length()

yes, I meant that if this is about bucket creation, we won't have any unicode character inside the string, am I correct?

balamurugana replied:
* In java, its required to set encoding to convert string into byte array, else you will get build error.
* In minio-java, `StandardCharsets.UTF_8` is used transparently ie S3 Rest API methods work with objects than strings/byte arrays.

vadmeste commented:
> @vadmeste The fix should be at https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/MinioClient.java#L2543. Instead of sending "", we would need to send new byte[0]

We also need to send content-length in complete multipart POST request. The reason of this patch is to send Content-Length header required for some proxies in POST requests.

vadmeste commented:
> Improvement to existing createRequest() is it work only with body as null, InputStream, RandomAccessFile, []byte than any object. This way length should be taken as is in RequestBody.contentLength(). As I mentioned previously this will lead to md5sum calculation for any non-null body.

Not sure about your exact meaning here, could you ping me when you are available ?

harshavardhana commented on api/src/main/java/io/minio/MinioClient.java:
```diff
@@ -1040,9 +1032,16 @@ private HttpResponse executePost(String bucketName, String objectName, Map<Strin
            InternalException {
     updateRegionCache(bucketName);
+    int len = -1;
+    if (data instanceof String) {
```
This type of special handling is not a good idea..  There is some issue here.

harshavardhana commented:
Is there any pending here? @vadmeste

vadmeste commented:
> Is there any pending here? @vadmeste

@harshavardhana, I addressed some issues mentioned in comments but it looks that we need more change to simplify the overall architecture, @balamurugana had some ideas about simplification so I'll let him decide if that change should be done in a separate PR or not.

balamurugana commented:
> @harshavardhana, I addressed some issues mentioned in comments but it looks that we need more change to simplify the overall architecture, @balamurugana had some ideas about simplification so I'll let him decide if that change should be done in a separate PR or not.

@vadmeste Any other changes need to be a separate PR.

harshavardhana commented:
what is pending in this PR? @vadmeste @balamurugana

vadmeste commented:
@harshavardhana, this PR is ready

balamurugana commented:
> @harshavardhana, this PR is ready

@vadmeste This PR hasn't been addressed the review comment.  Let me send you a PR to your fork with those changes.

balamurugana commented:
> @harshavardhana, this PR is ready
> 
> @vadmeste This PR hasn't been addressed the review comment. Let me send you a PR to your fork with those changes.

I sent a PR https://github.com/vadmeste/minio-java/pull/1 to your fork.  Please review and merge it accordingly.

balamurugana commented:
@vadmeste Please rebase
